Add user-level documentation for built-in timers (#2064)
nest/nest-simulator: doc/userdoc/guides/built-in_timers.rst

Built-in timers
===============

Basic timers
------------

Basic built-in timers keep track of the time NEST spent for network construction and actual simulation (propagation of the network state). These timers are active in all simulations with NEST, and the measured times can be queried using ``GetKernelStatus``. For example:

::

    nest.GetKernelStatus('time_simulate')

The following basic time measurements are available:

+-----------------------------+----------------------------------+
|Name                         |Explanation                       |
+=============================+==================================+

> |``time_construction_create`` |Cumulative time NEST spent        |

Reviewer: Will it be clear to users what is meant by "cumulative"? E.g. a note like

"Cumulative measurements are collected from the start of NEST or the most recent call to ``ResetKernel``."

|			      |creating	neurons and devices      |
+-----------------------------+----------------------------------+
|``time_construction_connect``|Cumulative time NEST spent        |
|                             |creating connections              |
+-----------------------------+----------------------------------+
|``time_simulate``            |Time NEST spent in the last       |
|                             |``Simulate()``                    |
+-----------------------------+----------------------------------+

.. note ::

>    After network construction in preparation of the actual simulation, NEST needs to build the pre-synaptic part of the connection infrastructure, which requires MPI communication (`Jordan et al. 2018 <https://doi.org/10.3389/fninf.2018.00002>`__). This happens only for the first call to ``Simulate()`` unless connectivity changed meanwhile, and it may cause significant overhead adding to ``time_simulate``. Therefore, the cumulative time NEST spent for building the pre-synaptic connection infrastructure is also tracked by a basic timer and available in the kernel dictionary as ``time_communicate_prepare``.

Reviewer: Wouldn't it be better to include ``time_communicate_prepare`` also in the table above, to reduce the risk that it will be overlooked and then explain it in the note?

>    After network construction in preparation of the actual simulation, NEST needs to build the pre-synaptic part of the connection infrastructure, which requires MPI communication (`Jordan et al. 2018 <https://doi.org/10.3389/fninf.2018.00002>`__). This happens only for the first call to ``Simulate()`` unless connectivity changed meanwhile, and it may cause significant overhead adding to ``time_simulate``. Therefore, the cumulative time NEST spent for building the pre-synaptic connection infrastructure is also tracked by a basic timer and available in the kernel dictionary as ``time_communicate_prepare``.

Reviewer 2: ```suggestion
   While preparing the actual simulation after network construction, NEST needs to build the pre-synaptic part of the connection infrastructure, which requires MPI communication (`Jordan et al. 2018 <https://doi.org/10.3389/fninf.2018.00002>`__). This happens only for the first call to ``Simulate()`` unless connectivity changed in the meantime, and it may cause significant overhead by adding to ``time_simulate``. Therefore, the cumulative time NEST spent for building the pre-synaptic connection infrastructure is also tracked by a basic timer and available in the kernel dictionary as ``time_communicate_prepare``.
```

In the context of NEST performance monitoring, other useful kernel-dictionary items are:

+-----------------------+----------------------------------+
|Name                   |Explanation                       |
+=======================+==================================+
|``biological_time``    |Cumulative simulated time         |
+-----------------------+----------------------------------+

> |``local_spike_counter``|Number of spikes emitted by the   |

Reviewer: Just noticed that the existing doc on this counter in PyNEST https://github.com/nest/nest-simulator/blob/2080ce33d9578df092ad296a3a49c0ab69d00742/pynest/nest/lib/hl_api_simulation.py#L363-L367 is wrong (also for 2.20.1). I will create a PR.

Author: Thanks!

|                       |neurons represented on this MPI   |
|			|rank during the last              |
|                       |``Simulate()``                    |
+-----------------------+----------------------------------+

.. note ::

   ``nest.ResetKernel()`` resets all time measurements as well as ``biological_time`` and ``local_spike_counter``.

Detailed timers
---------------

> Detailed built-in timers can be activated (and again deactivated) prior to compilation through the cmake flag ``with-detailed-timers``. They provide further insights into the time NEST spends in different phases of the simulation cycle, but they can impact the runtime. Therefore, detailed timers are by default inactive.

Reviewer: ```suggestion
Detailed built-in timers can be activated (and again deactivated) prior to compilation through the cmake flag ``-Dwith-detailed-timers``. They provide further insights into the time NEST spends in different phases of the simulation cycle, but they can impact the runtime. Therefore, detailed timers are by default inactive.
```

> Detailed built-in timers can be activated (and again deactivated) prior to compilation through the cmake flag ``with-detailed-timers``. They provide further insights into the time NEST spends in different phases of the simulation cycle, but they can impact the runtime. Therefore, detailed timers are by default inactive.

Reviewer 2: ```suggestion
Detailed built-in timers can be activated (and again deactivated) prior to compilation through the cmake flag ``with-detailed-timers``. They provide further insights into the time NEST spends in different phases of the simulation cycle, but they can impact the runtime. Therefore, detailed timers are inactive by default.
```
